### PR TITLE
Update LanguageServerHost to use NamedPipeServerStream

### PR DESCRIPTION
**Pull request by beccamc**
Current communication between client and the Rolsyn language server is writing to STDOUT. This makes the connection vulnerable to any code in the process writing to stdout stream and corrupting the data.

We have seen at least one bug with this problem: https://github.com/microsoft/vscode-dotnettools/issues/333

The fix it to use a named pipe to communicate. 
- Client (e.g. VS Code) will kick off this server process
- Create the NamedPipeServerStream. Specify [CurrentUserOnly flag ](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipes.pipeoptions?view=net-7.0), **which indicates that the pipe can only be connected to a client created by the same user.**  
- Pass the name of the pipe back to the client, who will connect 
- When the client connects, send the stream objects to Json-RPC

A note about names... 
- Unix-type systems use a file name to create the connection. 
- Windows systems use a pipe name. The format for the name is different in .NET code and Nodejs code. 

This has been tested on both windows and mac.

### CODE REVIEW

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs [hunk 1]
```diff
@@ -96,11 +99,22 @@ static async Task RunAsync(ServerConfiguration serverConfiguration, Cancellation
...
 
-    var server = new LanguageServerHost(Console.OpenStandardInput(), Console.OpenStandardOutput(), exportProvider, loggerFactory.CreateLogger(nameof(LanguageServerHost)));
+    var languageServerLogger = loggerFactory.CreateLogger(nameof(LanguageServerHost));
+
+    var (clientPipeName, serverPipeName) = CreateNewPipeNames();
+    var pipeServer = new NamedPipeServerStream(serverPipeName, PipeDirection.InOut, 1, PipeTransmissionMode.Byte, PipeOptions.CurrentUserOnly | PipeOptions.Asynchronous);
+
+    // Send the named pipe connection info to the client 
+    var clientPipeTransmissionInformation = new KeyValuePair<string, string>(NamedPipeKey, clientPipeName);
+    languageServerLogger.LogInformation(JsonConvert.SerializeObject(clientPipeTransmissionInformation));
```
Comment (dibarbet) [inline src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs:112]: I think we should explicitly write this to stdout - `LogInformation` might not always go exactly where we want it in the future.

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs [hunk 2]
```diff
@@ -96,11 +99,22 @@ static async Task RunAsync(ServerConfiguration serverConfiguration, Cancellation
...
     Microsoft.CodeAnalysis.EditAndContinue.EditAndContinueMethodDebugInfoReader.IgnoreCaseWhenComparingDocumentNames = Path.DirectorySeparatorChar == '\\';
 
-    var server = new LanguageServerHost(Console.OpenStandardInput(), Console.OpenStandardOutput(), exportProvider, loggerFactory.CreateLogger(nameof(LanguageServerHost)));
+    var languageServerLogger = loggerFactory.CreateLogger(nameof(LanguageServerHost));
+
+    var (clientPipeName, serverPipeName) = CreateNewPipeNames();
+    var pipeServer = new NamedPipeServerStream(serverPipeName, PipeDirection.InOut, 1, PipeTransmissionMode.Byte, PipeOptions.CurrentUserOnly | PipeOptions.Asynchronous);
+
+    // Send the named pipe connection info to the client 
+    var clientPipeTransmissionInformation = new KeyValuePair<string, string>(NamedPipeKey, clientPipeName);
```
Comment (dibarbet) [inline src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs:111]: consider creating a simple record type like this - https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Protocol/Handler/Testing/TestProgress.cs#L10 instead

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs [hunk 3]
```diff
@@ -96,11 +99,22 @@ static async Task RunAsync(ServerConfiguration serverConfiguration, Cancellation
...
     StarredCompletionAssemblyHelper.InitializeInstance(serverConfiguration.StarredCompletionsPath, loggerFactory, serviceBrokerFactory);
-
     // TODO: Remove, the path should match exactly. Workaround for https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1830914.
     Microsoft.CodeAnalysis.EditAndContinue.EditAndContinueMethodDebugInfoReader.IgnoreCaseWhenComparingDocumentNames = Path.DirectorySeparatorChar == '\\';
 
-    var server = new LanguageServerHost(Console.OpenStandardInput(), Console.OpenStandardOutput(), exportProvider, loggerFactory.CreateLogger(nameof(LanguageServerHost)));
+    var languageServerLogger = loggerFactory.CreateLogger(nameof(LanguageServerHost));
+
+    var (clientPipeName, serverPipeName) = CreateNewPipeNames();
+    var pipeServer = new NamedPipeServerStream(serverPipeName, PipeDirection.InOut, 1, PipeTransmissionMode.Byte, PipeOptions.CurrentUserOnly | PipeOptions.Asynchronous);
```
Comment (dibarbet) [inline src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs:108]: what is the `1` for?  Could use a named param

**Review by dibarbet — COMMENTED** [review]
mostly looks good!  a couple of hopefully simple comments

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs (hunk 2 above)
Comment (beccamc) [reply src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs]: I put it in `src/Features/LanguageServer/Protocol/Handler/`. Is that an appropriate location?

Comment (dibarbet) [reply src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs]: Next to the Program file is probably better, but thats more of a nit.

**Review by dibarbet — APPROVED** [review]

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs (hunk 2 above)
Comment (beccamc) [reply src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs]: Sure, I can move it

**Pull request merged**